drivers/soundwire/master.c

```diff
@@ -154,7 +154,12 @@ int sdw_master_device_add(struct sdw_bus *bus, struct device *parent,
 	bus->dev = &md->dev;
 	bus->md = md;
 
+	pm_runtime_set_autosuspend_delay(&bus->md->dev, 3000);
+	pm_runtime_use_autosuspend(&bus->md->dev);
+	pm_runtime_mark_last_busy(&bus->md->dev);
+	pm_runtime_set_active(&bus->md->dev);
 	pm_runtime_enable(&bus->md->dev);
+	pm_runtime_idle(&bus->md->dev);
```

Comment:
do I understand it correctly, that with this patch the problem, described in the commit message, doesn't occur any more, because in those cases, when rt-PM is trying to activate a child, now the parent isn't yet suspended because of the autosuspend delay?.. If that's correct, it doesn't seem right to me, relying on autosuspend to keep the parent on while we (re-)activate a child?

Reply:
no, the point is that we let the parent be part of the pm_runtime flow. 
When we activate a child, the parent will be activated by construction. Likewise, when the child is no longer active, the parent will be suspended as well.

This initial pm_runtime_idle is just to make sure the device becomes suspended IF its children are also suspended. Otherwise we've set it as active initially and it remains active forever.

Reply:
Sorry, my question wasn't to that `pm_runtime_idle()` call but to the complete code block, rather to the autosuspend handling.  Whether setting parent's autosuspend timeout we make sure it's still active in the beginning of the runtime suspend process?
BTW, in the commit message you probably meant "Force the parent device"

Reply:
The commit message says, that this patch fixes the
> runtime PM trying to activate child device sdw:3:25d:715:0 but parent (sdw-master-3) is not active

error message by "forcing the parent device to be active and use autosuspend to avoid these warnings." That makes me think, that without this patch the parent has already suspended when the driver is trying to activate the child. And with this patch the parent isn't yet suspended at that time, because it's still in its autosuspend delay?

Reply:
For PM runtime, a child can only become active when its parents are also active, and conversely a parent can only suspend when all of its children are suspended.

This sdw-master-N parent is an artificial layer we don't really needed, so we are just making it 'transparent' from the pm_runtime perspective, so that the codec device and the Intel master interface are connected and pm_runtime behaves as expected.

The first part is to make sure the device becomes active initially. If you don't no one will ever request the device to be active, and you get those warnings. By setting it to active, you allow the children to become active as well.

The autosuspend is the dual operation. You allow this device to suspend when no children are active.

Reply:
@plbossart so the error message appears not because the parent got suspended but because it wasn't activated to begin with? Got it, thanks!